### PR TITLE
Drop support for document interfaces that `extends Document`

### PR DESCRIPTION
**Pull request by vkarpov15**
**Summary**

Supporting document interfaces that `extends Document` in TypeScript is at best slow and bug prone, and at worst wrong. Given a type that `extends Document`, stripping out just the keys that are from Document is not feasible, which makes it hard for Mongoose to get an accurate representation of what fields are ok in a query filter or update - we don't want autocomplete to show `$locals`, virtuals, or other document-specific properties that aren't stored in MongoDB.

This PR removes `LeanDocument`, etc. and all the other workarounds we have to try to infer what the raw document type is. Instead, we assume that the first generic param to `Schema<>` and `Model<>` is the raw document type, with no virtuals, no methods, no `$locals`, etc. Just the data as it is stored in MongoDB.

This gets a little messy with subdocuments if you want to use subdocument methods. Right now we don't do a very good job of inferring what the hydrated document type should be from the schema. But we now support a `THydratedDocumentType` generic parameter to `Schema<>` and `Model<>` that contains the fully hydrated document type, which makes it easier to overwrite the hydrated document type. Below is an example.

```ts
  interface Nested {
    dummy?: string;
  }

  interface Parent {
    username?: string;
    map?: Map<string, string>;
    nested?: Nested;
    nestedArray?: Nested[];
  }

  // This is the hydrated document type
  type ParentDocument = HydratedDocument<Parent, {
    nested: HydratedSingleSubdocument<Nested>,
    nestedArray: HydratedArraySubdocument<Nested>[]
  }>;

  const NestedSchema = new Schema({
    dummy: { type: String }
  });

  type ParentModelType = Model<Parent, {}, {}, {}, ParentDocument>; // <-- set hydrated document type here

  const ParentSchema = new Schema<
  Parent,
  ParentModelType,
  {},
  {},
  {},
  {},
  DefaultSchemaOptions,
  Parent,
  ParentDocument // <-- and here
  >({
    username: { type: String },
    map: { type: Map, of: String },
    nested: { type: NestedSchema },
    nestedArray: [{ type: NestedSchema }]
  });

  const ParentModel = model<Parent, ParentModelType>('Parent', ParentSchema);

  {
    const doc = new ParentModel({
      username: 'user1',
      map: { key1: 'value1', key2: 'value2' },
      nested: { dummy: 'hello' },
      nestedArray: [{ dummy: 'hello again' }]
    });

    expectType<ParentDocument>(doc);
    expectType<Map<string, string> | undefined>(doc.map);
    doc.nested!.parent(); // works!
    doc.nestedArray?.[0].parentArray(); // works!
  }
```

A bit more friction than I'd like, but this opens up the possibility of automatically inferring the hydrated document type from the schema, which would be excellent. The downside is that inferring the hydrated document type is a bit risky, we've run into annoying edge cases in the past when we've tried to be a bit too clever with TypeScript. So I'm thinking we leave this workaround for now, and consider automatic inference for the future.

**Examples**

### CODE REVIEW

**Comment by hasezoey** [comment]
> Instead, we assume that the first generic param to Schema<> and Model<> is the raw document type, with no virtuals, no methods, no $locals, etc. Just the data as it is stored in MongoDB.

will this change impact things where monolithic types are used, for example a class (from typegoose)? 

> ```ts
>   type ParentDocument = HydratedDocument<Parent, {
>    nested: HydratedSingleSubdocument<Nested>,
>    nestedArray: HydratedArraySubdocument<Nested>[]
>  }>;
> ```

as a side note, what would be the best way to process such monolithic types's sub-documents into proper documents without having to need the user to manually specify everything?

**Review by hasezoey — COMMENTED** [review]
mostly looks good to me, though with some questions. (including the above about classes from something like typegoose)

also if i understand it correctly, the whole `Lean*` types have been removed and require the `DocType` to be the lean type directly? 
maybe the generic name should be renamed from `T` and `DocType` to something more representative, like `RawType` or `LeanType`

File: types/index.d.ts [hunk 1]
```diff
@@ -129,11 +130,12 @@ declare module 'mongoose' {
     ? IfAny<U, T & { _id: Types.ObjectId }, T & Required<{ _id: U }>>
     : T & { _id: Types.ObjectId };
 
-  export type HydratedDocument<DocType, TMethodsAndOverrides = {}, TVirtuals = {}> = DocType extends Document ? Require_id<DocType> : (Document<unknown, any, DocType> & Require_id<DocType> & TVirtuals & TMethodsAndOverrides);
+  export type HydratedDocument<DocType, TOverrides = {}> = (Document<unknown, any, DocType> & Require_id<DocType> & TOverrides);
+  export type HydratedSingleSubdocument<DocType, TOverrides = {}> = (Types.Subdocument<unknown> & Require_id<DocType> & TOverrides);
+  export type HydratedArraySubdocument<DocType, TOverrides = {}> = (Types.ArraySubdocument<unknown> & Require_id<DocType> & TOverrides);
```
[the remark this answers is not in the document]
Comment (vkarpov15) [reply types/index.d.ts:135]: No benefit. I'll remove those.

**Comment by hasezoey** [comment]
> Although there is a runtime error, not sure if that is expected or not.

the runtime error is probably because you ran it without the tsconfig option `emitDecoratorMetadata`

but yes, i will try this pr, if i can properly set it up

**Comment by hasezoey** [comment]
i just tested it, and it compiled fine after removing the references to `LeanDocument`

> and it compiles fine once I adjust for HydratedDocument now taking 2 arguments instead of 3

where did you need to change anything?

---

for reference i tested the following (after solving the mentioned expected errors above):
- `yarn run build` to see if the library itself is building
- `yarn run build:tests` to see if the tests for the library itself are building
- `yarn run test:types` to check if there are any type changes
- `yarn run test` to test runtime, which also works as expected, with only one change regarding discriminators (not related to this PR)

changes i had to do can be found at https://github.com/typegoose/typegoose/tree/mongoose70test

<details>
<summary>Steps to test the mentioned repository</summary>

- clone & enter the repository (at the mentioned branch)
- go to your mongoose installation
- run `yarn link`
- ensure you are on the correct mongoose branch
- go back to typegoose
- run `yarn link mongoose` to link mongoose
- run `yarn install` to install all dependencies (if done before the link, it should be re-done)
- now run the mentioned build / test commands

</details>

---

i kinda expected more problems, but it seems like there were actually none, though i also did not a "full-try" on a full-on project using typegoose, only typegoose and it's tests. it also seems like aside from no build errors, the usability of the types still is correct in my quick testing

**Comment by vkarpov15** [comment]
"where did you need to change anything?" <-- I got the following errors:

```
node_modules/@typegoose/typegoose/lib/hooks.d.ts:35:69 - error TS2707: Generic type 'HydratedDocument' requires between 1 and 2 type arguments.
```

So I just modified this PR to do:

```ts
export type HydratedDocument<DocType, TOverrides = {}, Unused = {}> = Document<unknown, any, DocType> & Require_id<DocType> & TOverrides;
```

--------

Thanks for taking a look! You're right that the primary change here is the return type for `toObject()` and `toJSON()`. Ideally most users won't be affected by this change, the idea is to make it explicit that the first generic param is the raw doc type, and also provide an override for the hydrated doc type. Currently, the first generic param is sometimes the raw doc type, but we also sorta support using it as the hydrated doc type in some cases, while also using LeanDocument to try to guess the raw doc type.

Re: "as a side note, what would be the best way to process such monolithic types's sub-documents into proper documents without having to need the user to manually specify everything?", we would have to add some logic to infer `THydratedDocumentType` from the schema. That will require some substantial work and be a bit risky, I'd prefer to defer that to a future release because this PR is big enough.

**Comment by birkskyum** [comment]
I've upgraded to v7, but I had to add the extends Document again, because otherwise all the typescript interfaces lack the .id / ._id properties.

**Comment by vkarpov15** [comment]
@birkskyum can you please provide a code sample that demonstrates the issue you're seeing?

**Comment by birkskyum** [comment]
@vkarpov15 , sure here is a repro of my issue, which has to do with recursiveness.
https://stackblitz.com/edit/typescript-zx4any?file=models/postings.ts

**Comment by hasezoey** [comment]
for reference [test.ts](https://github.com/typegoose/typegoose-testing/blob/mongoose12924birkskyum/src/test.ts), because the stackblitz didnt work for me

[full repro repository](https://github.com/typegoose/typegoose-testing/tree/mongoose12924birkskyum)

problem is on [this line](https://github.com/typegoose/typegoose-testing/blob/mongoose12924birkskyum/src/test.ts#L56)

because of the [direct reference to another interface](https://github.com/typegoose/typegoose-testing/blob/mongoose12924birkskyum/src/test.ts#L7) and mongoose does not change that to be a document (and from what i know, will never do, because it cant distinguish between subdocument and reference in this case, unless `PopulatedDoc` is used)

**Comment by birkskyum** [comment]
I understand, but if I use the hydrated type:

export type FlowDocument = ReturnType<typeof Flow['hydrate']>;

Inside of the interface isntead:

```ts
import { model, Schema } from 'mongoose';
import { FlowDocument, IFlowSchema } from './flow';

export interface ISpeciesSchema {
  flows: [FlowDocument];
  price: number;
}
```

I get the error that `'flows' is referenced directly or indirectly in its own type annotation.(2502)`

**Comment by birkskyum** [comment]
And with `exports Document` I don't have either of these two issues.

**Comment by hasezoey** [comment]
i dont know if this is how mongoose encourages to use it, but i personally would recommend to try changing `flows: [FlowDocument];` to `flows: [mongoose.HydratedDocument<FlowDocument>];` instead of a always `extends Document`

though this does not translate well for thing like `toObject` or `lean` types, because there typescript will still says that the type is a `HydratedDocument`

**Comment by birkskyum** [comment]
It seems like using  `flows: [HydratedDocument<IFlowSchema>];` does the trick

**Comment by vkarpov15** [comment]
@hasezoey it looks like [your repro](https://github.com/typegoose/typegoose-testing/tree/mongoose12924birkskyum) compiles as expected, at least on my end 

```
$ npm run build

> @typegoose/verify-typegoose-issues@0.0.1 build
> tsc

$ 
```

Is there something I need to change to get that test to fail?

**Comment by hasezoey** [comment]
> Is there something I need to change to get that test to fail?

did you check-out the correct branch? because running it locally it still gives me the errors:
```txt
$ yarn run build
$ tsc
node_modules/typescript/lib/lib.es5.d.ts:1584:11 - error TS2313: Type parameter 'P' has a circular constraint.

1584     [P in K]: T[P];
               ~

  src/test.ts:22:17
    22 const Species = mongoose.model('Species', speciesSchema);
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Circularity originates in type at this location.

src/test.ts:56:61 - error TS2339: Property 'id' does not exist on type 'IFlowSchema'.

56   console.log('species[0].flows[0].id', species[0].flows[0].id); // Property 'id' does not exist on type 'IFlowSchema'.ts(2339)
                                                               ~~


Found 2 errors in 2 files.

Errors  Files
     1  node_modules/typescript/lib/lib.es5.d.ts:1584
     1  src/test.ts:56
```

**Comment by vkarpov15** [comment]
@hasezoey you're right, I didn't check out the correct branch :man_facepalming: confirmed, I'll look into this

**Pull request merged**